### ISSUE TITLE
incompatible_strict_argument_ordering

### ISSUE THREAD

**Issue by laurentlb** [bazelbuild/bazel#6611]
Based on the discussion on https://github.com/bazelbuild/starlark/issues/13, we're going to make argument ordering stricter. It also makes the semantics simpler to understand, and can simplify tooling.

All arguments should be in this order:
- First, positional arguments
- Then, keyword arguments
- Then, an optional `*arg`
- Finally, an optional `**kwarg`

e.g. `f(a, b, c = 1, d = 2, *e, **f)`

Migration: update `buildifier` to automatically fix the call sites.
It is also trivial to fix manually: change the order of the arguments.

```diff
- f(a, b, *c, d = 2)
+ f(a, b, d = 2, *c)
```

Expected timeline:
 - released with Bazel 0.21 (December 2018), off by default
 - enabled by default with Bazel 0.23 (February 2019)

**Comment by laurentlb** [comment]
With this flag, only one downstream project fails: Gerrit.

```
ERROR: /var/lib/buildkite-agent/builds/bk-worker-ubuntu1604-java8-9b8n/bazel-downstream-projects/gerrit/tools/bzl/js.bzl:429:25: keyword argument is misplaced (keyword arguments must be before any *arg or **kwarg)
```

from https://gerrit.googlesource.com/gerrit/+/HEAD/tools/bzl/js.bzl#429
(cc @davido)

**Comment by davido** [comment]
cc @davidpersehouse.

I fixed that and many other issue in 2 months ago, but the Review was staled:

https://gerrit-review.googlesource.com/c/gerrit/+/205653/1/tools/bzl/js.bzl#429

because new buildifier release doesn't work on macOS.

**Comment by dpursehouse** [comment]
Here's the relevant PR on homebrew-core:  https://github.com/Homebrew/homebrew-core/pull/34815

**Comment by dpursehouse** [comment]
Note that the PR is using 0.19.2.1.  I have not tried it with the latest version 0.20.

**Comment by davido** [comment]
@dpursehouse Thanks for the pointer. What is the way forward? On Linux, I install buildifier from source. Would it be an option for you as well on macOS to move forward with fixing Gerrit build tool chain? We don't have a lot time, because I would really like to avoid the situation when latest Bazel releases cannot build Gerrit master (and the tip of supported stable branches).

**Comment by dpursehouse** [comment]
I wasn't able to build it from source either.  I'll try again tomorrow with the latest release.

**Comment by laurentlb** [comment]
The PR looks great. Can it be merged?
Using Buildifier feels like a separate issue (it shouldn't block the fix).

**Comment by davido** [comment]
> I wasn't able to build it from source either.

My alias command to build buildifier is here:

```
alias buildbuildifier='bazel build --workspace_status_command=`pwd`/status.sh buildifier:buildifier'
```

Used bazel version is: 

```
Build label: 0.22.0rc2
```

And most recent buildtool master just works: http://paste.openstack.org/show/740696

```
$ buildbuildifier
[...]
external/com_google_protobuf: warning: directory does not exist.
Target //buildifier:buildifier up-to-date:
  bazel-bin/buildifier/linux_amd64_stripped/buildifier
INFO: Elapsed time: 125.923s, Critical Path: 18.72s
INFO: 281 processes: 280 linux-sandbox, 1 local.
INFO: Build completed successfully, 295 total actions
```

Test:

```
$ bazel-bin/buildifier/linux_amd64_stripped/buildifier --version
buildifier version: 0.20.0 
buildifier scm revision: bf564b4925ab5876a3f64d8b90fab7f769013d42
```

**Comment by dpursehouse** [comment]
> The PR looks great. Can it be merged?
> Using Buildifier feels like a separate issue (it shouldn't block the fix).

Go ahead.  We'll deal with the issue in Gerrit.  Sorry for causing delays.

**Comment by dpursehouse** [comment]
New PR on homebrew to update to 0.20.0 building it with bazel instead of go: https://github.com/Homebrew/homebrew-core/pull/35816

**Comment by davido** [comment]
@dpursehouse I see that `buildifier` PR to upgrade to 0.20.0 was merged in Homebrew, but unfortunately `buildifier`'s `--lint=fix` options seems to be broken in 0.20.0 release. I filed this issue https://github.com/bazelbuild/buildtools/issues/505 upstream and provided a reproducer, basically issue this command in Gerrit Code Review tree and compare the difference between buildifier 0.19.2 (works) and 0.20.0 (broken):

```bash
find . \( -name BUILD -o -name "*.bzl" \) -print xargs buildifier --lint=fix
```

**Comment by dpursehouse** [comment]
@davido yes, I overlooked your comment before I went ahead and upgraded to 0.20.0

Do we then need to wait for a fixed version, or just use 0.20.0 without making use of the `--lint` feature?

**Comment by philwo** [comment]
I verified that this flag was flipped and thus will be included in the Bazel 0.23.0 release.
